### PR TITLE
SystemVerilog stubs generation

### PR DESCRIPTION
**Pull request by Muxianesty**
This Pull Request introduces a new option for Utopia HLS CLI to generate stub modules for SystemVerilog, which can be useful in SystemVerilog syntax parsing (e.g. with Verilator).

### CODE REVIEW

**Comment by Muxianesty** [comment]
**The next commit should fix the existing CI to build CTemplate as well.**

File: src/model/dfcir/lib/dfcir/conversions/FIRRTLStubsGeneratorPass.cpp [hunk 1]
```diff
@@ -0,0 +1,164 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of the Utopia HLS Project, under the Apache License v2.0
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2021-2024 ISP RAS (http://www.ispras.ru)
```
Comment (ssmolov) [inline src/model/dfcir/lib/dfcir/conversions/FIRRTLStubsGeneratorPass.cpp:5]: 2021-2024 -> 2024

File: src/model/dfcir/lib/dfcir/conversions/FIRRTLStubsGeneratorPass.cpp [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+                                       unsigned latency) {
+    TemplateDictionary *result = dict->AddSectionDictionary("BINARY_MODULES");
+    auto moduleName = module.getModuleName();
+    if (moduleName.contains(ADD_MODULE)) {
+      result->SetValue("OP", "+");
+    } else if (moduleName.contains(SUB_MODULE)) {
+      result->SetValue("OP", "-");
+    } else {
+      result->SetValue("OP", "*");
+    }
```
Comment (ssmolov) [inline src/model/dfcir/lib/dfcir/conversions/FIRRTLStubsGeneratorPass.cpp:62]: Add separate branch for MUL operation, report an error ("Unsupported module: " ?) at the default branch

File: src/model/dfcir/lib/dfcir/conversions/FIRRTLStubsGeneratorPass.cpp [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+    result->SetFormattedValue("REPEAT3", "%d", repeat3);
+    int32_t diff = std::min(width3, rWidth);
+    result->SetFormattedValue("DIFF", "%d", diff - 1);
+    int32_t cat = std::max(width3, rWidth);
+    result->SetFormattedValue("CAT", "%d", cat - 1);
+    if (isSigned) {
+        result->SetFormattedValue("REPEAT_VAL1", "%s[%d]", arg1, width1 - 1);
+        result->SetFormattedValue("REPEAT_VAL2", "%s[%d]", arg2, width2 - 1);
+        result->SetFormattedValue("REPEAT_VAL3", "r[%d][%d]",
+                                  latency - 1, rWidth - 1);
```
Comment (ssmolov) [inline src/model/dfcir/lib/dfcir/conversions/FIRRTLStubsGeneratorPass.cpp:96]: indent

File: src/model/dfcir/lib/dfcir/conversions/FIRRTLStubsGeneratorPass.cpp [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+    } else {
+        result->SetValue("REPEAT_VAL1", "1'h0");
+        result->SetValue("REPEAT_VAL2", "1'h0");
+        result->SetValue("REPEAT_VAL3", "1'h0");
```
Comment (ssmolov) [inline src/model/dfcir/lib/dfcir/conversions/FIRRTLStubsGeneratorPass.cpp:100]: indent

File: src/model/dfcxx/include/dfcxx/typedefs.h [hunk 5]
```diff
@@ -62,6 +62,7 @@ enum Scheduler {
 // Used for accessing specified output format paths.
 enum class OutputFormatID : uint8_t {
   SystemVerilog = 0,
+  SVStubs,
```
Comment (ssmolov) [inline src/model/dfcxx/include/dfcxx/typedefs.h:65]: probably renaming?

File: src/options.h [hunk 6]
```diff
@@ -36,6 +36,7 @@
 #define ASAP_SCHEDULER_JSON "asap_scheduler"
 #define LP_SCHEDULER_JSON "lp_scheduler"
 #define OUT_SV_JSON "out_sv"
+#define OUT_SV_STUBS_JSON "out_sv_stubs"
```
Comment (ssmolov) [inline src/options.h:39]: rename

File: src/options.h [hunk 7]
```diff
@@ -49,6 +50,7 @@
 #define LP_SCHEDULER_FLAG CLI_FLAG("l")
 #define OUTPUT_GROUP "output"
 #define OUT_SV_ARG CLI_ARG("out-sv")
+#define OUT_SV_STUBS_ARG CLI_ARG("out-sv-stubs")
```
Comment (ssmolov) [inline src/options.h:53]: rename

File: src/options.h [hunk 8]
```diff
@@ -184,6 +186,9 @@ struct HlsOptions final : public AppOptions {
     outputGroup->add_option(OUT_SV_ARG,
                             outNames[OUT_FORMAT_ID_INT(SystemVerilog)],
                             "Path to output the SystemVerilog module");
+    outputGroup->add_option(OUT_SV_STUBS_ARG,
+                            outNames[OUT_FORMAT_ID_INT(SVStubs)],
+                            "Path to output stubs for SystemVerilog modules");
```
Comment (ssmolov) [inline src/options.h:191]: rename

File: src/options.h [hunk 9]
```diff
@@ -198,6 +203,7 @@ struct HlsOptions final : public AppOptions {
     get(json, ASAP_SCHEDULER_JSON, asapScheduler);
     get(json, LP_SCHEDULER_JSON,   lpScheduler);
     get(json, OUT_SV_JSON,         outNames[OUT_FORMAT_ID_INT(SystemVerilog)]);
+    get(json, OUT_SV_STUBS_JSON,   outNames[OUT_FORMAT_ID_INT(SVStubs)]);
```
Comment (ssmolov) [inline src/options.h:206]: rename

File: config.json [hunk 10]
```diff
@@ -4,6 +4,7 @@
     "asap_scheduler" : false,
     "lp_scheduler"   : false,
     "out_sv"         : "",
+    "out_sv_stubs"   : "",
```
[the remark this answers is not in the document]
Comment (Muxianesty) [reply config.json:7]: Fixed.

File: src/model/dfcir/lib/dfcir/conversions/FIRRTLStubsGeneratorPass.cpp (hunk 1 above)
Comment (Muxianesty) [reply src/model/dfcir/lib/dfcir/conversions/FIRRTLStubsGeneratorPass.cpp:5]: Fixed here and in `templates/stubs.tpl`.

File: src/model/dfcir/lib/dfcir/conversions/FIRRTLStubsGeneratorPass.cpp (hunk 3 above)
Comment (Muxianesty) [reply src/model/dfcir/lib/dfcir/conversions/FIRRTLStubsGeneratorPass.cpp:96]: Fixed.

File: src/model/dfcir/lib/dfcir/conversions/FIRRTLStubsGeneratorPass.cpp (hunk 4 above)
Comment (Muxianesty) [reply src/model/dfcir/lib/dfcir/conversions/FIRRTLStubsGeneratorPass.cpp:100]: Fixed.

File: src/model/dfcxx/include/dfcxx/typedefs.h (hunk 5 above)
Comment (Muxianesty) [reply src/model/dfcxx/include/dfcxx/typedefs.h:65]: Fixed.

**Review by Muxianesty — COMMENTED** [review]
Fixed according to comments.

**Review by ssmolov — APPROVED** [review]
ok!

**Pull request merged**